node/network/bridge/src/lib.rs

```diff
@@ -0,0 +1,909 @@
+	/// A view update from a peer.
+	#[codec(index = "2")]
+	ViewUpdate(View),
+}
+
+/// Information about the notifications protocol. Should be used during network configuration
+/// or shortly after startup to register the protocol with the network service.
+pub fn notifications_protocol_info() -> (ConsensusEngineId, std::borrow::Cow<'static, [u8]>) {
+	(POLKADOT_ENGINE_ID, POLKADOT_PROTOCOL_NAME.into())
+}
+
+/// An action to be carried out by the network.
+#[derive(PartialEq)]
+pub enum NetworkAction {
+	/// Note a change in reputation for a peer.
+	ReputationChange(PeerId, ReputationChange),
+	/// Write a notification to a given peer.
+	WriteNotification(PeerId, Vec<u8>),
+}
+
+/// An abstraction over networking for the purposes of this subsystem.
+pub trait Network: Send + 'static {
+	/// Get a stream of all events occurring on the network. This may include events unrelated
+	/// to the Polkadot protocol - the user of this function should filter only for events related
+	/// to the [`POLKADOT_ENGINE_ID`](POLKADOT_ENGINE_ID).
+	fn event_stream(&mut self) -> BoxStream<'static, NetworkEvent>;
+
+	/// Get access to an underlying sink for all network actions.
+	fn action_sink<'a>(&'a mut self) -> Pin<
+		Box<dyn Sink<NetworkAction, Error = SubsystemError> + Send + 'a>
+	>;
+
+	/// Report a given peer as either beneficial (+) or costly (-) according to the given scalar.
+	fn report_peer(&mut self, who: PeerId, cost_benefit: ReputationChange)
+		-> BoxFuture<SubsystemResult<()>>
+	{
+		async move {
+			self.action_sink().send(NetworkAction::ReputationChange(who, cost_benefit)).await
+		}.boxed()
+	}
+
+	/// Write a notification to a peer on the [`POLKADOT_ENGINE_ID`](POLKADOT_ENGINE_ID) topic.
+	fn write_notification(&mut self, who: PeerId, message: Vec<u8>)
+		-> BoxFuture<SubsystemResult<()>>
+	{
+		async move {
+			self.action_sink().send(NetworkAction::WriteNotification(who, message)).await
+		}.boxed()
+	}
+}
+
+impl Network for Arc<sc_network::NetworkService<Block, Hash>> {
+	fn event_stream(&mut self) -> BoxStream<'static, NetworkEvent> {
+		sc_network::NetworkService::event_stream(self, "polkadot-network-bridge").boxed()
+	}
+
+	fn action_sink<'a>(&'a mut self)
+		-> Pin<Box<dyn Sink<NetworkAction, Error = SubsystemError> + Send + 'a>>
+	{
+		use futures::task::{Poll, Context};
+
+		// wrapper around a NetworkService to make it act like a sink.
+		struct ActionSink<'b>(&'b sc_network::NetworkService<Block, Hash>);
+
+		impl<'b> Sink<NetworkAction> for ActionSink<'b> {
+			type Error = SubsystemError;
+
+			fn poll_ready(self: Pin<&mut Self>, _: &mut Context) -> Poll<SubsystemResult<()>> {
+				Poll::Ready(Ok(()))
+			}
+
+			fn start_send(self: Pin<&mut Self>, action: NetworkAction) -> SubsystemResult<()> {
+				match action {
+					NetworkAction::ReputationChange(peer, cost_benefit) => self.0.report_peer(
+						peer,
+						cost_benefit,
+					),
+					NetworkAction::WriteNotification(peer, message) => self.0.write_notification(
```

Comment:
[`sc_network::NetworkService::write_notification`](https://github.com/paritytech/substrate/blob/master/client/network/src/service.rs#L555) is a synchronous method:

```rust
	pub fn write_notification(&self, target: PeerId, engine_id: ConsensusEngineId, message: Vec<u8>) {
```

Thus it does not need to be awaited.

How can it be a synchronous method even though the underlying network logic (libp2p) operates in an async style?

Because it is using an [unbounded channel](https://github.com/paritytech/substrate/blob/master/client/network/src/service.rs#L556).

Why is `Network::write_notification` an `async` trait method (returning a future) even though the concrete implementation is not?

To allow `sc_network::NetworkService::write_notification` to eventually be `async` and thus exercise proper backpressure across the system.

@drahnr does this explanation answer your question?

Reply:
Misread `NetworkService` as `Network` *sigh*